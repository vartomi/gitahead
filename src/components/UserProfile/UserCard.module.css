.container {
    padding: 24px;
    display: flex;
    border-bottom: 1px solid #ddd;
}

.container:hover {
    cursor: pointer;
}

.container:last-child {
    border-bottom: none;
}

.container:hover {
    background-color: #eee;
}

.img {
    max-width: 90px;
}

.imgContainer {
    border-radius: 90px;
    height: 90px;
    min-width: 90px;
    overflow: hidden;
}

.profileContainer {
    margin-left: 16px;
    display: flex;
    flex-grow: 1;
    flex-direction: column;
}

.userHeader > span {
    margin-right: 8px;
}

.userMain {
    flex-grow: 1;
}

.userFooter {
    display: flex;
    justify-content: flex-end;
}

.userFooter > span {
    text-transform: uppercase;
    color: #aaa;
    margin-left: 8px;
}

.userFooter > .value {
    color: #444;
}