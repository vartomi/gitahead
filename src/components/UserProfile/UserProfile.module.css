.container {
    width: 600px;
    margin-top: 48px;
    border: 1px solid #eee;
    box-shadow: 0px 5px 10px #ccc;    
    background-color: white;
    border-bottom: none;
}

.innerContainer {
    padding: 48px;
}

.imgContainer {
    text-align: center;
}

.img {
    max-width: 180px;
    border-radius: 180px;
}

.userHeader {
    text-align: center;
    margin-bottom: 32px;
}

.userName {
    font-size: 24px;
}

.userMain {
    margin-bottom: 32px;
    text-align: center;
}

.userMain > div {
    display: flex;
    justify-content: center;
    align-items: center;
}

.userMain > div svg {
    margin-right: 4px;
    fill: #444;
}

.userFooter {
    display: flex;
    justify-content: center;
}

.userFooter > span {
    text-transform: uppercase;
    color: #aaa;
    margin-left: 8px;
}

.userFooter > .value {
    color: #444;
}

.buttonLink {
    display: block;
    width: 100%;
    padding: 16px;
    text-align: center;
    background-color: #282c34;
    color: white;
    text-decoration: none;
}

.buttonLink:hover {
    background-color: #484e5a;
    text-decoration: none;
}