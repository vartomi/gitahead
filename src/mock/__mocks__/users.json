[
    {
        "login": "test-login",
        "id": 1,
        "node_id": "test",
        "avatar_url": "test-avatar",
        "gravatar_id": "",
        "url": "test",
        "html_url": "test",
        "followers_url": "test/followers",
        "following_url": "test/following{/other_user}",
        "gists_url": "test/gists{/gist_id}",
        "starred_url": "test/starred{/owner}{/repo}",
        "subscriptions_url": "test/subscriptions",
        "organizations_url": "test/orgs",
        "repos_url": "test/repos",
        "events_url": "test/events{/privacy}",
        "received_events_url": "test/received_events",
        "type": "User",
        "site_admin": false,
        "name": "Test Data",
        "company": "test",
        "blog": null,
        "location": null,
        "email": null,
        "hireable": null,
        "bio": "",
        "public_repos": 0,
        "public_gists": 0,
        "followers": 0,
        "following": 0,
        "created_at": "2009-08-01T18:39:25Z",
        "updated_at": "2020-01-27T18:17:21Z"
    }
]