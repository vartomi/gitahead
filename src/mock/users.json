[
  {
    "login": "mojombo",
    "id": 1,
    "node_id": "MDQ6VXNlcjE=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/1?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/mojombo",
    "html_url": "https://github.com/mojombo",
    "followers_url": "https://api.github.com/users/mojombo/followers",
    "following_url": "https://api.github.com/users/mojombo/following{/other_user}",
    "gists_url": "https://api.github.com/users/mojombo/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/mojombo/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mojombo/subscriptions",
    "organizations_url": "https://api.github.com/users/mojombo/orgs",
    "repos_url": "https://api.github.com/users/mojombo/repos",
    "events_url": "https://api.github.com/users/mojombo/events{/privacy}",
    "received_events_url": "https://api.github.com/users/mojombo/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Tom Preston-Werner",
    "company": null,
    "blog": "http://tom.preston-werner.com",
    "location": "San Francisco",
    "email": null,
    "hireable": null,
    "bio": null,
    "public_repos": 61,
    "public_gists": 62,
    "followers": 21770,
    "following": 11,
    "created_at": "2007-10-20T05:24:19Z",
    "updated_at": "2020-01-18T05:35:17Z"
  },
  {
    "login": "defunkt",
    "id": 2,
    "node_id": "MDQ6VXNlcjI=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/2?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/defunkt",
    "html_url": "https://github.com/defunkt",
    "followers_url": "https://api.github.com/users/defunkt/followers",
    "following_url": "https://api.github.com/users/defunkt/following{/other_user}",
    "gists_url": "https://api.github.com/users/defunkt/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/defunkt/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/defunkt/subscriptions",
    "organizations_url": "https://api.github.com/users/defunkt/orgs",
    "repos_url": "https://api.github.com/users/defunkt/repos",
    "events_url": "https://api.github.com/users/defunkt/events{/privacy}",
    "received_events_url": "https://api.github.com/users/defunkt/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Chris Wanstrath",
    "company": null,
    "blog": "http://chriswanstrath.com/",
    "location": null,
    "email": null,
    "hireable": null,
    "bio": "üçî",
    "public_repos": 107,
    "public_gists": 273,
    "followers": 20886,
    "following": 210,
    "created_at": "2007-10-20T05:24:19Z",
    "updated_at": "2019-11-01T21:56:00Z"
  },
  {
    "login": "addyosmani",
    "id": 110953,
    "node_id": "MDQ6VXNlcjExMDk1Mw==",
    "avatar_url": "https://avatars2.githubusercontent.com/u/110953?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addyosmani",
    "html_url": "https://github.com/addyosmani",
    "followers_url": "https://api.github.com/users/addyosmani/followers",
    "following_url": "https://api.github.com/users/addyosmani/following{/other_user}",
    "gists_url": "https://api.github.com/users/addyosmani/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addyosmani/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addyosmani/subscriptions",
    "organizations_url": "https://api.github.com/users/addyosmani/orgs",
    "repos_url": "https://api.github.com/users/addyosmani/repos",
    "events_url": "https://api.github.com/users/addyosmani/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addyosmani/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy Osmani",
    "company": "Google",
    "blog": "https://www.addyosmani.com",
    "location": "Mountain View, California",
    "email": null,
    "hireable": null,
    "bio": "Engineering Manager at Google working on Chrome & Web Platform",
    "public_repos": 307,
    "public_gists": 172,
    "followers": 34886,
    "following": 254,
    "created_at": "2009-08-01T18:39:25Z",
    "updated_at": "2020-01-27T18:17:21Z"
  },
  {
    "login": "addy",
    "id": 4543783,
    "node_id": "MDQ6VXNlcjQ1NDM3ODM=",
    "avatar_url": "https://avatars1.githubusercontent.com/u/4543783?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addy",
    "html_url": "https://github.com/addy",
    "followers_url": "https://api.github.com/users/addy/followers",
    "following_url": "https://api.github.com/users/addy/following{/other_user}",
    "gists_url": "https://api.github.com/users/addy/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addy/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addy/subscriptions",
    "organizations_url": "https://api.github.com/users/addy/orgs",
    "repos_url": "https://api.github.com/users/addy/repos",
    "events_url": "https://api.github.com/users/addy/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addy/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addison Shaw",
    "company": "Cerner",
    "blog": "https://addy.sh/",
    "location": "Kansas City, Missouri",
    "email": null,
    "hireable": null,
    "bio": "Software Engineer @ Cerner Corporation",
    "public_repos": 32,
    "public_gists": 13,
    "followers": 37,
    "following": 53,
    "created_at": "2013-05-27T22:57:16Z",
    "updated_at": "2020-01-18T15:27:02Z"
  },
  {
    "login": "addykim",
    "id": 7654513,
    "node_id": "MDQ6VXNlcjc2NTQ1MTM=",
    "avatar_url": "https://avatars0.githubusercontent.com/u/7654513?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/addykim",
    "html_url": "https://github.com/addykim",
    "followers_url": "https://api.github.com/users/addykim/followers",
    "following_url": "https://api.github.com/users/addykim/following{/other_user}",
    "gists_url": "https://api.github.com/users/addykim/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/addykim/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/addykim/subscriptions",
    "organizations_url": "https://api.github.com/users/addykim/orgs",
    "repos_url": "https://api.github.com/users/addykim/repos",
    "events_url": "https://api.github.com/users/addykim/events{/privacy}",
    "received_events_url": "https://api.github.com/users/addykim/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy Kim",
    "company": "Ubiquity6",
    "blog": "",
    "location": "San Francisco, CA",
    "email": null,
    "hireable": true,
    "bio": null,
    "public_repos": 51,
    "public_gists": 1,
    "followers": 24,
    "following": 20,
    "created_at": "2014-05-21T06:24:55Z",
    "updated_at": "2019-12-09T18:57:13Z"
  },
  {
    "login": "nerdyaddy",
    "id": 4039883,
    "node_id": "MDQ6VXNlcjQwMzk4ODM=",
    "avatar_url": "https://avatars1.githubusercontent.com/u/4039883?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/nerdyaddy",
    "html_url": "https://github.com/nerdyaddy",
    "followers_url": "https://api.github.com/users/nerdyaddy/followers",
    "following_url": "https://api.github.com/users/nerdyaddy/following{/other_user}",
    "gists_url": "https://api.github.com/users/nerdyaddy/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/nerdyaddy/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/nerdyaddy/subscriptions",
    "organizations_url": "https://api.github.com/users/nerdyaddy/orgs",
    "repos_url": "https://api.github.com/users/nerdyaddy/repos",
    "events_url": "https://api.github.com/users/nerdyaddy/events{/privacy}",
    "received_events_url": "https://api.github.com/users/nerdyaddy/received_events",
    "type": "User",
    "site_admin": false,
    "name": "Addy",
    "company": null,
    "blog": "",
    "location": "Mumbai",
    "email": null,
    "hireable": null,
    "bio": null,
    "public_repos": 4,
    "public_gists": 0,
    "followers": 3,
    "following": 0,
    "created_at": "2013-04-02T18:11:39Z",
    "updated_at": "2018-01-22T12:44:25Z"
  }
]