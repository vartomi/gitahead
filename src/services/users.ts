import mockUserList from '../mock/users.json';
import { GITHUB_API_USERS, GITHUB_QUERY_PAGE_USERS, SIMULATED_DELAY } from '../configs/constants';

export interface User {
    id: number;
    login: string;
    avatar: string;
    url: string;
    name: string;
    blog: string | null;
    bio: string | null;
    email: string | null;
    company: string | null;
    location: string | null;
    nrOfRepositories: number;
    nrOfFollowers: number;
    nrOfFollowing: number;
    joinedAt: string;
}

export const mapRawToUser: (rawData: any) => User = function ({
    id,
    avatar_url,
    login,
    name,
    bio,
    blog,
    email,
    public_repos,
    followers,
    following,
    created_at,
    html_url,
    company,
    location
}): User {
    return {
        id,
        login,
        avatar: avatar_url,
        url: html_url,
        name,
        blog,
        bio,
        email,
        location,
        company,
        nrOfRepositories: public_repos,
        nrOfFollowers: followers,
        nrOfFollowing: following,
        joinedAt: created_at
    };
}

let users: User[] = [];

export const getRawData = async function (link: string): Promise<[]> {
    const response: Response = await fetch(link);
    if (!response.ok) {
        throw new Error(response.statusText);
    }

    let rawData: [] = await response.json();
    return rawData;
}


export const getUsers = async function (withApi?: boolean): Promise<User[]> {
    if (withApi) {
        let rawData: any = await getRawData(GITHUB_API_USERS)
        rawData = await Promise.all(rawData.map(async (data: any) => {
            const meta: any = await getRawData(data.url);
            return { ...data, ...meta };
        }));

        users = rawData.map(mapRawToUser);
        return users;
    } else {
        users = mockUserList.map(mapRawToUser);

        return new Promise((resolve, _) => {
            setTimeout(() => resolve(users), SIMULATED_DELAY);
        });
    }
}


export const findUsers = function (searchValue: string): Promise<User[]> {
    const lowerCaseSearchValue = searchValue.toLowerCase();
    return new Promise((resolve, _) => {
        const result: User[] = users.filter(user =>
            (user.login && user.login.toLowerCase().startsWith(lowerCaseSearchValue)) ||
            (user.name && user.name.toLowerCase().startsWith(lowerCaseSearchValue))
        ).slice(0, 5);
        setTimeout(() => resolve(result), SIMULATED_DELAY);
    });
};

export const searchOnGithub = function (searchValue: string): void {
    window.open(`${GITHUB_QUERY_PAGE_USERS}?q=${searchValue}&type=users`);
}