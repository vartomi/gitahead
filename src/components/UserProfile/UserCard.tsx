import React from 'react';
import styles from './UserCard.module.css';
import { User } from '../../services/users';

type UserCardProps = {
    user: User;
    onSelect: any;
}

const UserCard: React.FC<UserCardProps> = ({ user, onSelect }) => {
    const { login, name, avatar, bio, nrOfFollowers, nrOfRepositories, nrOfFollowing } = user;
    return (
        <div className={styles.container} onClick={() => onSelect(user)} tabIndex={0}>
            <div className={styles.imgContainer}><img src={avatar} className={styles.img} alt="avatar" /></div>
            <div className={styles.profileContainer}>
                <div className={styles.userHeader}>
                    <span><strong>{name}</strong></span>
                    <span>{login}</span>
                </div>
                <div className={styles.userMain}>{bio}</div>
                <div className={styles.userFooter}>
                    <span>Following </span><span className={styles.value}>{nrOfFollowing}</span>
                    <span>Followers </span><span className={styles.value}>{nrOfFollowers}</span>
                    <span>Repositories </span><span className={styles.value}>{nrOfRepositories}</span>
                </div>
            </div>
        </div >
    );
}

export default UserCard;
