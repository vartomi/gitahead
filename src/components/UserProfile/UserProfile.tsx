import React from 'react';
import styles from './UserProfile.module.css';
import { User } from '../../services/users';
import { ReactComponent as LocationIcon } from '../../assets/location.svg';
import { ReactComponent as BlogIcon } from '../../assets/blog.svg';
import { ReactComponent as CompanyIcon } from '../../assets/company.svg';

type UserProfileProps = {
    user: User;
}

const UserProfile: React.FC<UserProfileProps> = ({ user: { url, login, name, avatar, blog, bio, nrOfFollowers, nrOfRepositories, nrOfFollowing, company, location, joinedAt } }) => {
    return (
        <div className={styles.container}>
            <div className={styles.innerContainer}>
                <div className={styles.imgContainer}><img src={avatar} className={styles.img} alt="avatar" /></div>
                <div className={styles.profileContainer}>
                    <div className={styles.userHeader}>
                        <div className={styles.userName}><strong>{name}</strong></div>
                        <div>{login}</div>
                    </div>
                    <div className={styles.userMain}>
                        {company && <div><CompanyIcon />{company}</div>}
                        {location && <div><LocationIcon />{location}</div>}
                        {blog && <div><BlogIcon /><a href={blog}>{blog}</a></div>}
                        <p>{bio}</p>
                    </div>
                    <div className={styles.userFooter}>
                        <span>Following </span><span className={styles.value}>{nrOfFollowing}</span>
                        <span>Followers </span><span className={styles.value}>{nrOfFollowers}</span>
                        <span>Repositories </span><span className={styles.value}>{nrOfRepositories}</span>
                    </div>
                </div>
            </div>
            <a className={styles.buttonLink} href={url}>Open in GitHub</a>
        </div>
    );
}

export default UserProfile;
