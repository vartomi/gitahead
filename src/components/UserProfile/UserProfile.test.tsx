import { shallow } from 'enzyme'
import UserProfile from './UserProfile';
import React from 'react';
import { User } from '../../services/users';

const mockUser: User = {
    id: 1,
    name: 'name',
    login: 'login',
    avatar: 'avatar',
    url: 'url',
    blog: 'blog',
    email: 'email',
    bio: 'bio',
    company: 'company',
    location: 'location',
    joinedAt: 'date',
    nrOfFollowers: 0,
    nrOfFollowing: 0,
    nrOfRepositories: 0
};

describe('UserProfile component', () => {

    it('should render snapshot', () => {
        const wrapper = shallow(<UserProfile user={mockUser} />)
        expect(wrapper).toMatchSnapshot();
    });
});