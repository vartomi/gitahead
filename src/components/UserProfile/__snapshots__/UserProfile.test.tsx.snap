// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile component should render snapshot 1`] = `
<div
  className="container"
>
  <div
    className="innerContainer"
  >
    <div
      className="imgContainer"
    >
      <img
        alt="avatar"
        className="img"
        src="avatar"
      />
    </div>
    <div
      className="profileContainer"
    >
      <div
        className="userHeader"
      >
        <div
          className="userName"
        >
          <strong>
            name
          </strong>
        </div>
        <div>
          login
        </div>
      </div>
      <div
        className="userMain"
      >
        <div>
          <ForwardRef(SvgCompany) />
          company
        </div>
        <div>
          <ForwardRef(SvgLocation) />
          location
        </div>
        <div>
          <ForwardRef(SvgBlog) />
          <a
            href="blog"
          >
            blog
          </a>
        </div>
        <p>
          bio
        </p>
      </div>
      <div
        className="userFooter"
      >
        <span>
          Following 
        </span>
        <span
          className="value"
        >
          0
        </span>
        <span>
          Followers 
        </span>
        <span
          className="value"
        >
          0
        </span>
        <span>
          Repositories 
        </span>
        <span
          className="value"
        >
          0
        </span>
      </div>
    </div>
  </div>
  <a
    className="buttonLink"
    href="url"
  >
    Open in GitHub
  </a>
</div>
`;
